### ISSUE TITLE
Enhancement: Explicitly specify IVSHMEM device on host

### ISSUE THREAD

**Issue by tadly** [gnif/LookingGlass#154]
@martinellimarco Is doing amazing work over at [scream](https://github.com/duncanthrax/scream) and has brought us 8 channel support already (amazing with an isolated network).
Now he's working on using an IVSHMEM device (just like looking-glass) to further improve performance/reliability ([related issue](https://github.com/duncanthrax/scream/issues/36)).

While doing some tests we encountered that the device order is not a given and can change on subsequent boots.
In other words, having two or more IVSHMEM devices will cause looking-glass to fail (if it picks the wrong device).

We have not yet come up with a proper/elegant solution but [are brainstorming ideas](https://github.com/duncanthrax/scream/issues/36#issuecomment-484876168).

We would very much appreciate it if you'd chime in with ideas and maybe even a _fix_ for looking-glass.

**Comment by gnif** [comment]
Scream streams audio over a network which makes sense, but over IVSHMEM it's just silly.

If you really want to transfer audio directly to the host via shared memory and are willing to write a sound driver (as you are already), you should be instead writing a virtual audio device for qemu that suit's your project's requirements, just like has been done for VirtIO SCSI, Keyboard, Mouse, Network, etc.

**Comment by paul70078** [comment]
I've started working on configuring the used IVSHMEM device. I think this could be useful to use LG together with other IVSHMEM interfaces. This would also allow duplicating multiple displays

**Comment by paul70078** [comment]
I think this issue is not about if audio over IVSHMEM with scream is a proper solution or fits scream. it is about allowing communication of multiple devices over IVSHMEM regions without conflicts.

**Comment by tadly** [comment]
>...it is about allowing communication of multiple devices over IVSHMEM regions without conflicts

Yes. Though this only came about due to our (seemingly silly) task of using audio over IVSHMEM :)

It's good to know that this would be considered a hack as I actually thought it's a neat solution to the never ending story of _bad qemu audio_.

The more you know :)

**Comment by paul70078** [comment]
I've created a basic implementation at my fork [paul70078/LookingGlass](https://github.com/paul70078/LookingGlass).
Prebuilt binary: [looking-glass-host.exe](https://send.firefox.com/download/18594b6e63ca176c/#7MuX8BMwgjRfEFR-ZZNjsg)

I've added the commandline argument `-d ?` to list all available IVSHMEM devices and `-d <bus>,<slot>,<function>` with bus, slot and function are to be replaced by the decimal values of the device in the libvirt configuration.

I recommend also using the argument `-f` to view the commandline output and to easier kill the process. If the specified device isn't found the application hangs. If no device is specified it searches for an device with `bus=0x13`, `slot=0x0` and `function=0x0`.

Todo:

- [ ] Fix forwarding the device selection to the ICSHMEM-class to be less of a hack
- [ ] Exit application if device is not found
- [ ] Select first device if if no device was configured

I'm also thinking about making the `function` key optional because it should be `0` in most cases.

I'm not quite sure how to forward the selected device properly. One option would be to set the device index as member variable and changing it would Reinitialize the device. A static function would retrieve the device index from the bus, slot and function index.

**Comment by gnif** [comment]
Looks good but please note that the c++ host application has a limited lifespan and will be deprecated soon, it would be better to update the `c-host` under `platform/windows`

**Comment by paul70078** [comment]
I will port the changes in the next few days. It should be pretty easy to integrate the improvements from my previous comment :)

**Comment by paul70078** [comment]
I've integrated some changes, but have problem compiling it with mingw. I don't know how to link/include direct-x. I've tried using the includes from mingw-64 but they don't work for me.

**Comment by gnif** [comment]
The project already performs the required steps to link against dx11, ensure you're building using `cmake` in a proper bash terminal (I am using GitBash)

**Comment by paul70078** [comment]
I'm still missing dxgi.h (and maybe other dx11 header files too) :/

**Comment by gnif** [comment]
`dxgi.h` comes with the mingw package from winbuilds
the dx11 headers are not part of mingw and are produced by the cmake file at build time.

**Comment by khogeland** [comment]
@paul70078 Thanks for the fork! But, unfortunately, when attempting to configure qemu:

```
error: XML error: Invalid PCI address 0000:13:00.0. slot must be >= 1
```

**Comment by khogeland** [comment]
@paul70078 A small PR for you: https://github.com/paul70078/LookingGlass/pull/1

Looking Glass and Scream are working fine now! :smile:

**Comment by gnif** [comment]
@khogeland this is NOT the place to report issues with a fork of Looking Glass, your comments are to be removed.

**Comment by khogeland** [comment]
That was rude. Should you delete your own comments discussing the fork? Feel free to delete this one as well.

**Comment by gnif** [comment]
@khogeland it was off topic, if you need to report an issue with the forked code you need to open an issue on @paul70078's fork. The topic of this issue is to discuss merging this feature into Looking Glass, discussion about the fork along these lines is on topic.

**Comment by gnif** [comment]
I have reviewed and commented on the changes on your branch. I am going to close this issue as GitHub has a review feature as part of a Pull Requests that we should be using, please when you are ready issue a PR for your changes and I will have another look :).

Thank you for your efforts this far

**Comment by gnif** [comment]
The new option API is available for use, please see `dxgi.c` for an example of usage.
